src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java

```diff
@@ -671,225 +570,60 @@ private void addStoneRecipes(Consumer<IFinishedRecipe> consumer) {
-        .key('L', Tags.Items.PURPLE_SLIMEBALL)
-        .patternLine("#X#")
-        .patternLine("L L")
-        .patternLine(" L ")
-        .addCriterion("has_item", this.hasItem(Items.STRING))::build)
-      .setAdvancement(new ResourceLocation(TConstruct.modID, "recipes/tinkers_gadgets/slimesling/purple"), ConditionalAdvancement.builder()
-        .addCondition(new PulseLoadedCondition(TinkerPulseIds.TINKER_GADGETS_PULSE_ID))
-        .addAdvancement(Advancement.Builder.builder()
-          .withParentId(new ResourceLocation("recipes/root"))
-          .withRewards(AdvancementRewards.Builder.recipe(purpleSlimeSlingId))
-          .withCriterion("has_item", hasItem(Items.STRING))
-          .withCriterion("has_the_recipe", new RecipeUnlockedTrigger.Instance(purpleSlimeSlingId))
-          .withRequirementsStrategy(IRequirementsStrategy.OR))
-      ).build(consumer, purpleSlimeSlingId);
+        .addRecipe(ShapedRecipeBuilder.shapedRecipe(GadgetItems.slime_sling.get(slime))
+          .setGroup("tconstruct:slimesling")
+          .key('#', Items.STRING)
+          .key('X', WorldBlocks.congealed_slime.get(slime))
+          .key('L', slime.getSlimeBallTag())
+          .patternLine("#X#")
+          .patternLine("L L")
+          .patternLine(" L ")
+          .addCriterion("has_item", this.hasItem(Items.STRING))::build)
+        .setAdvancement(new ResourceLocation(TConstruct.modID, "recipes/tinkers_gadgets/slimesling/" + slime.getName()), ConditionalAdvancement.builder()
+          .addCondition(new PulseLoadedCondition(TinkerPulseIds.TINKER_GADGETS_PULSE_ID))
+          .addAdvancement(Advancement.Builder.builder()
+            .withParentId(new ResourceLocation("recipes/root"))
+            .withRewards(AdvancementRewards.Builder.recipe(slimeSlingId))
+            .withCriterion("has_item", hasItem(Items.STRING))
+            .withCriterion("has_the_recipe", new RecipeUnlockedTrigger.Instance(slimeSlingId))
+            .withRequirementsStrategy(IRequirementsStrategy.OR))
+        ).build(consumer, slimeSlingId);
+    }
   }
 
   private void addSlimeBootsRecipes(Consumer<IFinishedRecipe> consumer) {
-    ResourceLocation bloodSlimeBootsId = new ResourceLocation(TConstruct.modID, "gadgets/slimeboots/blood");
-    ConditionalRecipe.builder()
-      .addCondition(new PulseLoadedCondition(TinkerPulseIds.TINKER_GADGETS_PULSE_ID))
-      .addRecipe(ShapedRecipeBuilder.shapedRecipe(GadgetItems.slime_boots_blood)
-        .setGroup("tconstruct:slime_boots")
-        .key('#', WorldBlocks.congealed_blood_slime)
-        .key('X', Tags.Items.BLOOD_SLIMEBALL)
-        .patternLine("X X")
-        .patternLine("# #")
-        .addCriterion("has_item", this.hasItem(Items.SLIME_BALL))::build)
-      .setAdvancement(new ResourceLocation(TConstruct.modID, "recipes/tinkers_gadgets/slimeboots/blood"), ConditionalAdvancement.builder()
+    for (SlimeBlock.SlimeType slime : SlimeBlock.SlimeType.values()) {
```

Comment:
So, the thing special about pink slime boots is they can be crafted from any slime type. However, I think we are missing the recipe type for that recipe (prevents conflicts with the colored boots), so maybe leave this as pink for now and I will add it in a later commit.

Reply:
True, but you should also be able to produce a pink from a solid pink
craft. Perhaps after the loopy stuff make a specific pink_from_mixed
recipe manually.

Reply:
using tags, obviously. Our and vanilla slime's are all tagged under slimeballs, right?

Reply:
I had code in 1.12 that basically defined a list of "matches", so pink was actually handled through a fallback because pink lacked a recipe. I will look into the best way to remake it a bit, maybe just list the "unless" recipes would be cleaner

Reply:
Yeah. Well, we'll deal with that later.